oak/server/rust/oak_runtime/src/channel.rs

```diff
@@ -278,14 +335,34 @@ pub fn wait_on_channels(
     runtime: &RuntimeRef,
     readers: &[Option<&ChannelReader>],
 ) -> Vec<Result<bool, OakStatus>> {
+    let thread = platform::thread::current();
     while !runtime.is_terminating() {
+        // Create a new Arc each iteration to be dropped after `thread::park` e.g. when the thread
+        // is resumed. When the Arc is deallocated, any remaining `Weak` references in
+        // `Channel`s will be orphaned. This means thread::unpark will not be called multiple times.
+        // Even if thread unpark is called spuriously and we wake up early, no channel
+        // statuses will be ready and so we can just continue.
+        //
+        // Note we read statuses directly after adding waiters, before blocking to ensure that
+        // there are no messages, after we have been added as a waiter.
+
+        let thread_id = platform::thread::current().id();
+        let thread_ref = Arc::new(thread.clone());
+
+        for reader in readers {
+            if let Some(reader) = reader {
+                reader.add_waiter(thread_id, &thread_ref);
+            }
+        }
         let statuses = readers_statuses(readers);
 
         if statuses.iter().all(|s| s.is_err()) || statuses.iter().any(|s| s.unwrap_or(false)) {
             return statuses;
         }
 
-        crate::yield_thread();
+        println!("Statuses not ready, parking thread");
+
+        platform::thread::park();
```

Comment:
Might be helpful to have a `debug!` after wakeup.

Reply:
Done.